### PR TITLE
Custom commands or currency as rewards for time/kills/wins

### PR DESCRIPTION
**Pull request by abedaton**
I have added the possibility of executing commands for the 3 events (time, kills and wins), user can chose in the config if they just want a currency reward or commands:

>   time:
        enable: false
        # Do you want to reward the player with a currency ?
        isEconomy: false
        # If above is true, they will receive the amount bellow
        reward: 100
        # if isEconomy is false, enter a list of commands that should be executed instead (the / before is ignored)
        commands:
           - msg %name% Congratulation you survived !
           - /give %name% gold_ingot 1
        interval: 600

I added the string `EVENT_TIME_COMMANDS` if the player doesn't want any economy. The commands for the three events are stored in an `ArrayList<String>`.  If you find any bugs or want anything to change message me and I'll update the code !

### CODE REVIEW

File: src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java [hunk 1]
```diff
@@ -424,12 +433,22 @@ public void load(YamlFile cfg, @Nullable YamlFile storage){
 
 		// custom events
 		enableTimeEvent = cfg.getBoolean("custom-events.time.enable",false);
+		isEconomyTime = cfg.getBoolean("custom-events.time.isEconomy", false);
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java:436]: Change `isEconomy` to `is-economy`

File: src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java [hunk 2]
```diff
@@ -127,12 +127,21 @@
...
+
 	private boolean enableKillEvent;
+	private boolean isEconomyKill;
 	private double rewardKillEvent;
+	private ArrayList<String> killCommands;
+
 	private boolean enableWinEvent;
+	private boolean isEconomyWin;
 	private double rewardWinEnvent;
+	private ArrayList<String> winCommands;
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java:144]: Change array lists to lists

File: src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java [hunk 3]
```diff
@@ -424,12 +433,22 @@ public void load(YamlFile cfg, @Nullable YamlFile storage){
 
 		// custom events
 		enableTimeEvent = cfg.getBoolean("custom-events.time.enable",false);
+		isEconomyTime = cfg.getBoolean("custom-events.time.isEconomy", false);
 		rewardTimeEvent = cfg.getDouble("custom-events.time.reward",0);
+		timeCommands = new ArrayList<>(cfg.getStringList("custom-events.time.commands"));
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java:438]: No need to create a new array list from something that is already a list.

File: src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java [hunk 4]
```diff
@@ -999,4 +1018,31 @@ public void setTheEndUuid(String theEndUuid) {
 		this.theEndUuid = theEndUuid;
 	}
 
+
+	/********************/
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java:1022]: Whats the `/**********************/` for? If it's a division line please put some info in what its dividing.

File: src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java [hunk 5]
```diff
@@ -58,9 +60,21 @@ public void onPlayerDeath(PlayerDeathEvent event){
...
-				VaultManager.addMoney(killer, reward);
-				if(!Lang.EVENT_KILL_REWARD.isEmpty()){
-					killer.sendMessage(Lang.EVENT_KILL_REWARD.replace("%money%", ""+reward));
+				if (cfg.getIsEconomyKill()) {
+					VaultManager.addMoney(killer, reward);
+					if (!Lang.EVENT_KILL_REWARD.isEmpty()) {
+						killer.sendMessage(Lang.EVENT_KILL_REWARD.replace("%money%", "" + reward));
+					}
+				} else {
+					ArrayList<String> cmds = cfg.getKillCommands();
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java:69]: Change to a list.

File: src/main/java/com/gmail/val59000mc/players/PlayersManager.java [hunk 6]
```diff
@@ -824,9 +836,21 @@ public void killOfflineUhcPlayer(UhcPlayer uhcPlayer, @Nullable Location locatio
...
+						killer.sendMessage(Lang.EVENT_KILL_REWARD.replace("%money%", "" + reward));
+					}
+				} else {
+					ArrayList<String> cmds = cfg.getKillCommands();
+					if (cmds != null) {
+						for (String cmd : cmds) {
+							if (cmd.startsWith("/")) {
+								cmd = cmd.substring(1);
+							}
+							Bukkit.dispatchCommand(Bukkit.getConsoleSender(), cmd.replace("%name%", uhcKiller.getRealName()));
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/players/PlayersManager.java:851]: Put try catch around dispatching the command for `CommandException`.
https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/Bukkit.html#dispatchCommand-org.bukkit.command.CommandSender-java.lang.String-

File: src/main/resources/config.yml [hunk 7]
```diff
@@ -509,16 +509,35 @@ custom-events:
...
+      - msg %name% You got a kill !
+      - /give %name% diamond 5
   #Reward a price in your economy currency to the winners of the game
   win:
     enable: false
+    isEconomy: false
     reward: 1000
+    commands:
+      - msg %name% You won the game !
+      - /give %name% diamond 128
```
Comment (Mezy) [inline src/main/resources/config.yml:540]: Instead of adding a `isEconomy` option maybe execute both? And when reward is 0 or command list is empty that part doesn't get triggered.

File: src/main/resources/config.yml [hunk 8]
```diff
@@ -509,16 +509,35 @@ custom-events:
...
   time:
     enable: false
+    # Do you want to reward the player with a currency ?
+    isEconomy: false
+    # If above is true, they will receive the amount bellow
     reward: 100
+    # if isEconomy is false, enter a list of commands that should be executed instead (the / before is ignored)
+    commands:
+      - msg %name% Congratulation you survived !
+      - /give %name% gold_ingot 1
```
Comment (Mezy) [inline src/main/resources/config.yml:519]: I would either do all commands with / or without, this will bring confusion.

File: src/main/java/com/gmail/val59000mc/languages/Lang.java [hunk 9]
```diff
@@ -331,6 +332,7 @@ private void loadLangConfig() {
 
 		// Event
 		EVENT_TIME_REWARD = getString(lang, "event.time-reward", "&eYou have received %money% in your account for playing %time% , total playing time %totaltime%");
+		EVENT_TIME_COMMANDS = getString(lang, "event.time-commands", "&eYou successfully survived %time%, total playing time %totaltime%");
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/languages/Lang.java:335]: Why did you add this message for `EVENT_TIME` but not for `EVENT_KILL`? Also the message doesn't mention anything about the commands getting executed so I don't really see why it would be needed either. If people really want a message they can use a command for that.

File: src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java [hunk 10]
```diff
@@ -424,12 +433,22 @@ public void load(YamlFile cfg, @Nullable YamlFile storage){
...
+		isEconomyKill = cfg.getBoolean("custom-events.kill.isEconomy", false);
 		rewardKillEvent = cfg.getDouble("custom-events.kill.reward");
+		killCommands = new ArrayList<>(cfg.getStringList("custom-events.kill.commands"));
+
+
 		enableWinEvent = cfg.getBoolean("custom-events.win.enable",false);
+		isEconomyWin = cfg.getBoolean("custom-events.win.isEconomy", false);
 		rewardWinEnvent = cfg.getDouble("custom-events.win.reward",0);
+		winCommands = new ArrayList<>(cfg.getStringList("custom-events.win.commands"));
+
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/configuration/MainConfiguration.java:451]: Maybe loop trough the commands here to remove the initial / instead of redoing it every time the commands get executed.

File: src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java [hunk 11]
```diff
@@ -58,10 +62,21 @@ public void onPlayerDeath(PlayerDeathEvent event){
...
+					if (!Lang.EVENT_KILL_REWARD.isEmpty()) {
+						killer.sendMessage(Lang.EVENT_KILL_REWARD.replace("%money%", "" + reward));
+					}
 				}
+				// If the list is empty, this will never execute
+				killCommands.forEach(cmd -> {
+					try {
+						Bukkit.dispatchCommand(Bukkit.getConsoleSender(), cmd.replace("%name%", uhcKiller.getRealName()));
+					} catch (CommandException exception){
+						Bukkit.getLogger().warning("The command: '" + cmd + "' does not exists.");
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java:77]: This exception also gets thrown when there is an error in the command, so I'd change this message to:
```
Bukkit.getLogger().warning("[UhcCore] Failed to execute kill reward command: " + cmd);
```
And also always print exceptions to console so add
```
exception.printStackTrace();
```

File: src/main/java/com/gmail/val59000mc/players/PlayersManager.java [hunk 12]
```diff
@@ -426,19 +427,47 @@ public void setAllPlayersEndGame() {
 		Bukkit.getServer().getPluginManager().callEvent(event);
 
 		double reward = cfg.getRewardWinEnvent();
+		List<String> winCommands = cfg.getWinCommands();
+		boolean cmdsNeedsPlayer = false;
+		for (String cmd : winCommands){
+			if (cmd.contains("%name%")){
+				cmdsNeedsPlayer = true;
+				break;
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/players/PlayersManager.java:435]: Don't really understand this ``cmdsNeedsPlayer``, if one of the commands needs a play they all do?

File: src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java [hunk 13]
```diff
@@ -50,30 +62,51 @@ public void run() {
 		if(time%intervalTimeEvent == 0){
 			
 			if(enableTimeEvent){
-				
-				String message = Lang.EVENT_TIME_REWARD
+				String message;
+
+				message = Lang.EVENT_TIME_REWARD
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java:67]: Can be done in one line.

File: src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java [hunk 14]
```diff
@@ -50,30 +62,51 @@ public void run() {
...
+									Bukkit.dispatchCommand(Bukkit.getConsoleSender(), cmd.replace("%name%", uhcP.getRealName()));
+								} catch (CommandException exception) {
+									Bukkit.getLogger().warning("The command: '" + cmd + "' does not exists.");
+								}
+							});
+						}
+
+						if (!message.isEmpty()) {
 							p.sendMessage(message);
 						}
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java:90]: Why do you send the message regardless of money received here? That's different on the other events.

File: src/main/resources/config.yml [hunk 15]
```diff
@@ -509,16 +509,30 @@ custom-events:
   #Reward a player a price in your economy currency every time he is still alive after the interval (in seconds)
   time:
     enable: false
+    # Amount of money the player should get
```
Comment (Mezy) [inline src/main/resources/config.yml:512]: Add ``(Set to 0 for no money reward)``

File: src/main/resources/config.yml [hunk 16]
```diff
@@ -509,16 +509,30 @@ custom-events:
...
     reward: 100
+    # Enter a list of commands that should be executed instead
```
Comment (Mezy) [inline src/main/resources/config.yml:514]: Add ``(Change to commands: [] if you do not want any commands to be executed.)``

File: src/main/java/com/gmail/val59000mc/players/PlayersManager.java (hunk 12 above)
Comment (abedaton) [reply src/main/java/com/gmail/val59000mc/players/PlayersManager.java:435]: No need to run the command for winners.size() times if the command does not take the %name% placeholders, if it has the %name% placeholder, then it's because the command is specific to every player. If we run the command in the 
`for UhcPlayer player : winners`, then the command will be run a lot of times for nothing

Comment (Mezy) [reply src/main/java/com/gmail/val59000mc/players/PlayersManager.java:435]: I understand that but that's not what your code does. Your code checks if any of the commands contains %name% in which case it executes all commands for all winners.

Comment (Mezy) [reply src/main/java/com/gmail/val59000mc/players/PlayersManager.java:435]: You'll still have to remove the break. This way the winCommandsPlayer list will at most contain 1 command.

File: src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java [hunk 17]
```diff
@@ -58,10 +62,22 @@ public void onPlayerDeath(PlayerDeathEvent event){
...
-					killer.sendMessage(Lang.EVENT_KILL_REWARD.replace("%money%", ""+reward));
+				List<String> killCommands = cfg.getKillCommands();
+				if (reward > 0) {
+					VaultManager.addMoney(killer, reward);
+					if (!Lang.EVENT_KILL_REWARD.isEmpty()) {
+						killer.sendMessage(Lang.EVENT_KILL_REWARD.replace("%money%", "" + reward));
+					}
 				}
+				// If the list is empty, this will never execute
+				killCommands.forEach(cmd -> {
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java:73]: I'd change this to
```
cfg.getKillCommands().forEach(cmd -> {
```
And remove the unnecessary variable on line 65.

File: src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java [hunk 18]
```diff
@@ -37,6 +40,7 @@ public void onPlayerDeath(PlayerDeathEvent event){
 		MainConfiguration cfg = gm.getConfiguration();
 		UhcPlayer uhcPlayer = pm.getUhcPlayer(player);
 
+
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/listeners/PlayerDeathListener.java:43]: This return got in here somehow, it's not a big thing of course but it doesn't need to be there either.

File: src/main/java/com/gmail/val59000mc/players/PlayersManager.java [hunk 19]
```diff
@@ -426,19 +427,50 @@ public void setAllPlayersEndGame() {
...
+		}
+		winCommands.removeAll(winCommandsPlayer);
+
 		if(cfg.getEnableWinEvent()){
-			for(UhcPlayer player : winners){
+			for(UhcPlayer player : winners) {
 				try {
-					if(!Lang.EVENT_WIN_REWARD.isEmpty()){
-						player.getPlayer().sendMessage(Lang.EVENT_WIN_REWARD.replace("%money%", ""+reward));
+					if (reward != 0) {
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/players/PlayersManager.java:443]: Change to ``(reward > 0)``

File: src/main/java/com/gmail/val59000mc/players/PlayersManager.java [hunk 20]
```diff
@@ -426,19 +427,50 @@ public void setAllPlayersEndGame() {
...
+						if (!Lang.EVENT_WIN_REWARD.isEmpty()) {
+							player.getPlayer().sendMessage(Lang.EVENT_WIN_REWARD.replace("%money%", "" + reward));
+						}
+						VaultManager.addMoney(player.getPlayer(), reward);
+					}
+					if (!winCommandsPlayer.isEmpty()){
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/players/PlayersManager.java:449]: No need to check if that list is empty, if it is forEach won't do anything.

File: src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java [hunk 21]
```diff
@@ -50,30 +62,50 @@ public void run() {
...
 			if(enableTimeEvent){
-				
 				String message = Lang.EVENT_TIME_REWARD
 						.replace("%time%", TimeUtils.getFormattedTime(intervalTimeEvent))
 						.replace("%totaltime%", TimeUtils.getFormattedTime(time))
-						.replace("%money%", ""+reward);
-				
-				for(UhcPlayer uhcP : playingPlayers){
-					
+						.replace("%money%", "" + reward);
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java:68]: Move this part inside the ``if (reward > 0)`` statement.

File: src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java [hunk 22]
```diff
@@ -21,13 +24,22 @@
 	private boolean enableTimeEvent;
 	private long intervalTimeEvent;
 	private double reward;
+	private List<String> timeCommands;
+	private List<String> timeCommandsPlayers = new ArrayList<>();
```
Comment (Mezy) [inline src/main/java/com/gmail/val59000mc/threads/ElapsedTimeThread.java:28]: Instantiate ArrayList in constructor.
